### PR TITLE
Add package name to the protocol buffer definition

### PR DESCRIPTION
**Pull request by jenting**
Add package name to the protocol buffer definition file to avoid the namespace conflict.
Use the unique package name `longhorn.backingimagemanager.pkg.rpc`.

[Longhorn 3343](https://github.com/longhorn/longhorn/issues/3343)

Signed-off-by: JenTing Hsiao <jenting.hsiao@suse.com>

### CODE REVIEW

File: pkg/rpc/rpc.proto [hunk 1]
```diff
@@ -1,7 +1,11 @@
 syntax="proto3";
 
+package longhorn.io.backing.image.manager.pkg.rpc;
```
Comment (shuo-wu) [inline pkg/rpc/rpc.proto:3]: Too many layers... Should not be `longhorn.backingimagemanager.pkg.rpc`?

Comment (jenting) [reply pkg/rpc/rpc.proto]: Sounds good.

**Comment by shuo-wu** [comment]
BTW, have you tested the backward compatibility?

**Comment by jenting** [comment]
> BTW, have you tested the backward compatibility?

No, any idea on how to test it?

**Comment by shuo-wu** [comment]
> > BTW, have you tested the backward compatibility?
> 
> No, any idea on how to test it?

I would recommend testing with the instance manager. If everything works fine with the new instance manager, then you don't need to worry about this.

You can try the following:
1. Launch Longhorn system with old instance manager image.
2. Create and attach some volumes.
3. Upgrade the system with the new instance manager image.
4. Apply some operations for the volumes.
5. Do live upgrade for the volumes then retry applying some operations.

**Comment by jenting** [comment]
> BTW, have you tested the backward compatibility?

New longhorn manager is not backwards compatible with old backing image manager, as well as old instance manager, the controllers display the error messages
```shell
Error syncing Longhorn backing image manager longhorn-system/backing-image-manager-c00e-d5d8: BackingImageManagerController failed to sync longhorn-system/backing-image-manager-c00e-d5d8: failed to sync backing image manager pod: failed to get version: rpc error: code = Unimplemented desc = unknown service longhorn.backingimagemanager.pkg.rpc.BackingImageManagerService

error receiving next item in engine watch: rpc error: code = Unimplemented desc = unknown service longhorn.instancemanager.pkg.rpc.ProcessManagerService
```

**Comment by shuo-wu** [comment]
It suddenly occurred to me that we can retain the old version instance manager gRPC client in the latest longhorn manager repo. Then we can fall back to using the old client when encountering this kind of error. It's ugly but may work.

**Comment by jenting** [comment]
> It suddenly occurred to me that we can retain the old version instance manager gRPC client in the latest longhorn manager repo. Then we can fall back to using the old client when encountering this kind of error. It's ugly but may work.

I have an idea.

Since the current conflict is the **VersionResponse** message between [instance-manager](https://github.com/longhorn/longhorn-instance-manager/blob/884c12d4daede2abf1347f8153971f05f9313469/pkg/rpc/rpc.proto#L70) and [backing-image-manager](https://github.com/longhorn/backing-image-manager/blob/0103fcb66924708d0aa4ca3e64576711b3cced00/pkg/rpc/rpc.proto#L50).
And, the major backward compatibility concern is the instance-manager _only_ because the backing-image-manager will upgrade along with the longhorn-manager upgrade. Therefore, the solution now is we add the package name to the backing-image-manager _only_. Then, there would not have a message definition conflict between instance-manager and backing-image-manager anymore.

**Comment by shuo-wu** [comment]
Yes. The backing image manager compatibility is not a blocker here. 
But avoiding updating the instance manager is not a long-term solution... We still need to think about how we can fix it.

BTW, you need to bump up the version for backing image manager and its client (in longhorn manager).

**Comment by jenting** [comment]
> But avoiding updating the instance manager is not a long-term solution... We still need to think about how we can fix it.

Right, but fallback is not an easy way to maintain code because the `rpc.pb.go` is an auto-generated code.
I'd prefer we _do not_ update the instance manager at the moment, and for the future protocol buffer file, have a convention to always add a unique package name.

> BTW, you need to bump up the version for backing image manager and its client (in longhorn manager).

I have a draft PR in the longhorn-manager already, will update the version for backing image manager and its client once this PR be merged, thank you.

**Comment by shuo-wu** [comment]
In this PR, you need to update https://github.com/longhorn/backing-image-manager/blob/0103fcb66924708d0aa4ca3e64576711b3cced00/pkg/meta/version.go#L6-L7

**Review by shuo-wu — APPROVED** [review]

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
> I have an idea.
> 
> Since the current conflict is the **VersionResponse** message between [instance-manager](https://github.com/longhorn/longhorn-instance-manager/blob/884c12d4daede2abf1347f8153971f05f9313469/pkg/rpc/rpc.proto#L70) and [backing-image-manager](https://github.com/longhorn/backing-image-manager/blob/0103fcb66924708d0aa4ca3e64576711b3cced00/pkg/rpc/rpc.proto#L50). And, the major backward compatibility concern is the instance-manager _only_ because the backing-image-manager will upgrade along with the longhorn-manager upgrade. Therefore, the solution now is we add the package name to the backing-image-manager _only_. Then, there would not have a message definition conflict between instance-manager and backing-image-manager anymore.
> 
> I did a test and it's workable.

http://github.com/longhorn/longhorn-manager/blob/452ca6e313b7378c9bd4c3d35e1a66a6eadfbcde/controller/backing_image_controller.go#L298-L309

**Pull request merged**